set.mm

```diff
@@ -756151,6 +756179,534 @@ last digit of the integer part (for ` b = ` 10 the first digit before
+  simplest and earliest-discovered examples of a total computable function that
+  is not primitive recursive. ... One common version is the two-argument
+  Ackermann-P&eacute;ter function developed by R&oacute;zsa P&eacute;ter and
+  Raphael Robinson.  Its value grows very rapidly; for example, ` A ( 4 , 2 ) `
+  results in ` 2 ^ ; 6 5 5 3 6 - 3 `, an integer of 19,729 decimal digits."
+
+  In the following, the Ackermann function is defined as iterated 1-ary
+  function (also mentioned in Wikipedia), see ~ df-ack , based on a definition
+  ` IterComp ` of "the n-th iterate of (a class/function) f", see ~ df-itco .
+  As an illustration, we have ` ( ( IterComp `` F ) `` 3 ) `
+  ` = ( F o. ( F o. F ) ) ) ` (see ~ itcoval3 ).
+
+  The following recursive definition of the Ackermann function follows
+  immediately from the definition ~ df-ack : ` ( ( Ack `` ( M + 1 ) ) `` N ) `
+  ` = ( ( ( IterComp `` ( Ack `` M ) ) `` ( N + 1 ) ) `` 1 ) ) `.
+
+  That the definition ~ df-ack is equivalent to P&eacute;ter's definition is
+  proven by the following three theorems:
+
+  ~ ackval0val : ` ( ( Ack `` 0 ) `` M ) = ( M + 1 ) `
+  ~ ackvalsuc0val : ` ( ( Ack `` ( M + 1 ) ) `` 0 ) = ( ( Ack `` M ) `` 1 ) `
+  ~ ackvalsucsucval : ` ( ( Ack `` ( M + 1 ) ) `` ( N + 1 ) ) `
+                      ` = ( ( Ack `` M ) `` ( ( Ack `` ( M + 1 ) ) `` N ) ) `.
+
+  The initial values of the Ackermann function are calculated in the following
+  four theorems:
+
+  ~ ackval0012 : ` A ( 0 , 0 ) = 1 `,  ` A ( 0 , 1 ) = 2 `, ` A ( 0 , 2 ) = 3 `
+  ~ ackval1012 : ` A ( 1 , 0 ) = 2 `,  ` A ( 1 , 1 ) = 3 `, ` A ( 1 , 3 ) = 4 `
+  ~ ackval2012 : ` A ( 2 , 0 ) = 3 `,  ` A ( 2 , 1 ) = 5 `, ` A ( 2 , 3 ) = 7 `
+  ~ ackval3012 : ` A ( 3 , 0 ) = 5 `,  ` A ( 3 , 1 ) = ; 1 3 `,
+                 ` A ( 3 , 3 ) = ; 2 9 `
+
+$)
+
+  $c IterComp $.
+  $c Ack $.
+
+  $( Extend the definition of a class to include iterated functions. $)
+  citco $a class IterComp $.
+
+  $( Extend the definition of a class to include the Ackermann function
+     operator. $)
+  cack $a class Ack $.
+
+  ${
+    $d f g i j $.
+    $( Define a function (recursively) that returns the n-th iterate of a class
+       (usually a function) with regard to composition.  (Contributed by Tierry
+       Arnoux, 28-Apr-2024.)  (Revised by AV, 2-May-2024.) $)
+    df-itco $a |- IterComp = ( f e. _V
+         |-> seq 0 ( ( g e. _V , j e. _V |-> ( f o. g ) ) ,
+                     ( i e. NN0 |-> if ( i = 0 , ( _I |` dom f ) , f ) ) ) ) $.
+  $}
+
+  ${
+    $d f i j n $.
+    $( Define the Ackermann function (recursively).  (Contributed by Tierry
+       Arnoux, 28-Apr-2024.)  (Revised by AV, 2-May-2024.) $)
+    df-ack $a |- Ack = seq 0 ( ( f e. _V , j e. _V |-> ( n e. NN0
+                             |-> ( ( ( IterComp ` f ) ` ( n + 1 ) ) ` 1 ) ) ) ,
+                               ( i e. NN0 |-> if ( i = 0 ,
+                                        ( n e. NN0 |-> ( n + 1 ) ) , i ) ) ) $.
```

Comment:
Interestingly, the part `if ( i = 0 , ( n e. NN0 |-> ( n + 1 ) ) , i )` does not seem to be homogeneous:
* if `i = 0`, this is a function on `NN0`,
* otherwise this is the integer `i`

I think that "else" branch is actually not needed anywhere.

Reply:
> Interestingly, the part `if ( i = 0 , ( n e. NN0 |-> ( n + 1 ) ) , i )` does not seem to be homogeneous:
> 
>     * if `i = 0`, this is a function on `NN0`,
> 
>     * otherwise this is the integer `i`
> 
> 
> I think that "else" branch is actually not needed anywhere.

Yes, this is exactly the oddity I mentioned in my comment in the Google group.